HLRC Refactor out common client request methods

hub-cap opened the pull request:
The HLRC client served two purposes. It would encapsulate the client and
all communication to elasticsearch, as well as hold methods for all of
the top level core API calls. This commit removes all of the client and
communication and encapsulates it into RestRequestActions. This is now
passed into the subclients instead of the RestHighLevelClient itself.

hub-cap committed:
HLRC: reorg client package structure

The Request and Response classes are being moved out of protocol and in
to the HLRC. This commit gets the package structure ready for the move,
as well as moving all of the clients into the new package structure.

The main portion of this commit removes the low level guts of the HLRC
into a CoreClient, which then is given to every subclient to use. The
CoreClient holds the state of the HLRC, such as the actual RestClient
and NamedXContents.

elasticmachine commented:
Pinging @elastic/es-core-infra

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/CoreClient.java:
```diff
@@ -0,0 +1,508 @@
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Base client of
```
Doh, will fix!

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/CoreClient.java:
```diff
@@ -0,0 +1,508 @@
+/**
+ * Base client of
+ */
+public class CoreClient {
```
This is all just the bits of the request performing in the old high level client, moved into this class. There are no changes to this other than the copy/paste from the other client.

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -86,8 +86,8 @@
 
-final class RequestConverters {
-    static final XContentType REQUEST_BODY_CONTENT_TYPE = XContentType.JSON;
+public final class RequestConverters {
```
public was for the moved RequestConverters, which are now in their subpackages

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -1134,352 +1025,7 @@ public void putScriptAsync(PutStoredScriptRequest putStoredScriptRequest, Reques
+        coreClient.performRequestAsyncAndParseEntity(fieldCapabilitiesRequest, RequestConverters::fieldCaps, options,
             FieldCapabilitiesResponse::fromXContent, listener, emptySet());
     }
-
```
Starting here all of this was just moved into CoreClient

hub-cap commented:
Terribly sorry for the large PR. Its all mechanical, as the CoreClient is just a subset of methods that use to be on the RestHighLevelClient. Splitting them out caused a big diff. I also had to split the tests up to do the same type of thing, so that caused an even larger diff. :(

hub-cap committed:
Fix the commit msg

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/CoreClient.java:
```diff
@@ -0,0 +1,509 @@
+ * Base client in charge of interfacing with the low level {@link RestClient}. Every subclient should take a {@link CoreClient} in as its
+ * sole constructor argument, and use these methods to communicate with the server.
+ */
+public class CoreClient {
```
This is all just the bits of the request performing in the old high level client, moved into this class. There are no changes to this other than the copy/paste from the other client.

nik9000 replied:
Making it abstract and having the others extend it makes them all `Closeable` when really only this thing has any state. I like this way better because it lets us decide what "owns" the state.

I kind of like that the class is public and documented "for extensions". It makes it clear that we support extensions.

nik9000 replied:
I'm not particularly comfortable with the way that we add named writeables to the CoreClient but I do like not extending from it. I feel like extending from it is better implementation hiding than extending from it. Most users don't need to know about it so it is a good thing that it doesn't show up in the Javadoc of RestHighLevelClient at all, even in as "extends CoreClient".

hub-cap replied:
Also, extending from it coupled with moving the other clients means the methods have to be public. So then the HLRC has IDE completion on those methods as public methods when they want to only have the methods that dont have a namespace, and the namespaced sub-clients.

javanna replied:
good point @hub-cap , ok let's leave things as they are, that's the best we can do atm.

nik9000 commented:
Stick some havadov on RequestVonverters explaining that it is for
extensions and ship it before you get more merge conflicts. It is a great
step. I'm sure we'll want to iterate on it some more but let's do that in
follow-up changes. Thanks for doing this!

On Thu, Sep 20, 2018, 7:30 PM Michael Basnight <notifications@github.com>
wrote:

> *@hub-cap* commented on this pull request.
> ------------------------------
>
> In
> client/rest-high-level/src/main/java/org/elasticsearch/client/CoreClient.java
> <https://github.com/elastic/elasticsearch/pull/33926#discussion_r219348036>
> :
>
> > +import java.util.Map;
> +import java.util.Objects;
> +import java.util.Optional;
> +import java.util.ServiceLoader;
> +import java.util.Set;
> +import java.util.function.Function;
> +import java.util.stream.Collectors;
> +import java.util.stream.Stream;
> +
> +import static java.util.stream.Collectors.toList;
> +
> +/**
> + * Base client in charge of interfacing with the low level {@link RestClient}. Every subclient should take a {@link CoreClient} in as its
> + * sole constructor argument, and use these methods to communicate with the server.
> + */
> +public class CoreClient {
>
> This is all just the bits of the request performing in the old high level
> client, moved into this class. There are no changes to this other than the
> copy/paste from the other client.
>
> —
>
>
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/33926#pullrequestreview-157502918>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANLoqBalHCd_bqvfo7d7qGF-_fn0voqks5udCUhgaJpZM4WzM38>
> .
>

hub-cap committed:
Add some javadoc to Request Converters

javanna reviewed (changes requested):
I left some comments, nothing that will require super major rework though. I would probably mark this PR breaking-java and find a place to list what users have to do to migrate (update imports mainly?)

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -281,7 +166,7 @@ public final IndicesClient indices() {
      */
     public final ClusterClient cluster() {
-        return clusterClient;
+        return buildSubClient(ClusterClient::new);
```
It is a tiny object that'll die young so I figured this is fine.

javanna replied:
Is it not possible to have a singleton like before? All the clients could be created in the constructor? That would be better I guess

nik9000 replied:
Creating them in the ctor is fine with me. I don't think it is better but I don't think it is worse either.

javanna replied:
I don't get why you say it is not better. Users are most likely not going to do 

```
ClusterClient  clusterClient = client.cluster();
clusterClient.doSomething();
clusterClient.doSomethingElse();
```

but rather 

```
client.cluster().doSomething();
client.cluster().doSomethingElse();
```

and I don't think they would expect we'd create a new object each time. If we can avoid creating all those objects, we must be making things ever so slightly better by doing so, no? We are certainly not saving the world, but still...

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -86,8 +86,12 @@
+ * A set of methods used to facilitate conversion of specific actions requests into generic {@link Request}. These are meant to be helper
+ * methods used by subclients, which need to send generic {@link Request}'s down to the {@link RestClient}.
+ */
+public final class RequestConverters {
```
I'm torn here because extensions could really use these. But I'm not sure we're really ready to declare them stable. Basically the question is, "should we have to leave a breaking change doc every time we change one of these?"

nik9000 commented:
I think maybe we *should* pull out the `CoreClient` change from the package shuffling. We are debating exactly how to do that. I like this was and @javanna wants it another way. I think that change would be a lot smaller and less likely to have merge conflicts.

nik9000 commented:
I was suggesting doing the CoreClient change first and getting that one settled.

hub-cap commented:
im ++ to only doing the core client change since this the PR is merge conflict city already. Ill keep this PR open and issue a revert commit, and only pull whats relevant back in.

javanna reviewed (approved):
I am good with the direction, ship it. We do need to clarify tests on extension points so they reflect what we expect users to do. We can also do so as a follow-up.

hub-cap committed:
Revert "HLRC: reorg client package structure"

This reverts commit 99c0cfd1ab8d4ebd141637c9d8058437bbec4336.

hub-cap committed:
Revert "Fix the commit msg"

This reverts commit 39e3e24587b1895cb0256228d6cb0739311335f6.

hub-cap committed:
Revert "Add some javadoc to Request Converters"

This reverts commit 256b08047b65d9255017baaca6f6518d9841418a.

hub-cap committed:
HLRC Refactor out common client request methods

The HLRC client served two purposes. It would encapsulate the client and
all communication to elasticsearch, as well as hold methods for all of
the top level core API calls. This commit removes all of the client and
communication and encapsulates it into RestRequestActions. This is now
passed into the subclients instead of the RestHighLevelClient itself.

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestRequestActions.java:
```diff
@@ -0,0 +1,524 @@
+     * layer has been added to the ReST client, and requests should extend {@link Validatable} instead of {@link ActionRequest}.
+     */
+    @Deprecated
+    protected <Req extends ActionRequest, Resp> void performRequestAsync(Req request,
```
I think these should be public now?

hub-cap replied:
I will public'ize them once i move the package, subsequent PR to come.

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -247,23 +126,40 @@ protected RestHighLevelClient(RestClientBuilder restClientBuilder, List<NamedXCo
+    /**
+     * Returns the wrapper used to communicate with the low-level client that all subclients should be using.
+     */
+    public final RestRequestActions getRequestActions() {
```
I'd be more comfortable if this were protected or a public method like `<T> T buildExtension(Function<RestRequestActions, T>)`.

hub-cap replied:
In 054deea I dropped the visibility of `getRequestActions`, and added the `buildSubClient` method like you mention above.

hub-cap committed:
Fix visibility of the RestRequestActions from within HLRC

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Fix failing test

nik9000 commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:
```diff
@@ -199,29 +199,32 @@ public ActionRequestValidationException validate() {
         {
             ActionRequestValidationException actualException = expectThrows(ActionRequestValidationException.class,
-                    () -> restHighLevelClient.performRequest(request, null, RequestOptions.DEFAULT, null, null));
+                    () -> restHighLevelClient.getRequestActions().performRequest(request, null, RequestOptions.DEFAULT, null, null));
```
Should these be in a `RestRequestActionsTests`?

hub-cap replied:
good catch on splitting this up. incoming shortly. this also gets rid of a lot of uses of the `getRequestActions`, which leads me to wonder if we should have that method at all :)

hub-cap committed:
Split test

hub-cap commented:
@elasticmachine run gradle build tests

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Fix some compile errors

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
compile cleanup

hub-cap commented:
Seeing that so many ppl are merging HLRCs im going to revert the variable name change here so i stop risking compilation errors. Ill come back and rename them when all the APIs are done :)

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Revert the name change cuz muchos conflicts

hub-cap committed:
Add blank line

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap committed:
Merge remote-tracking branch 'upstream/master' into hlrc_reorg_final

hub-cap commented:
Closing as this never happened and I am awaiting further conversations on generating some of the code here
